src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionTelemetry.cs

```diff
@@ -0,0 +1,214 @@
+                    DisplayUnits = "ms"
+                };
+            }
+        }
+
+        private const int MaxIPFormattedLength = 128;
+
+        [Event(ResolutionStartEventId, Level = EventLevel.Informational)]
+        public ValueStopwatch ResolutionStart(string hostNameOrAddress)
+        {
+            Debug.Assert(hostNameOrAddress != null);
+
+            if (IsEnabled())
+            {
+                Interlocked.Increment(ref _lookupsRequested);
+                WriteEvent(ResolutionStartEventId, hostNameOrAddress);
+                return ValueStopwatch.StartNew();
+            }
+
+            return default;
+        }
+
+        [NonEvent]
+        public ValueStopwatch ResolutionStart(IPAddress address)
+        {
+            Debug.Assert(address != null);
+
+            if (IsEnabled())
+            {
+                Interlocked.Increment(ref _lookupsRequested);
+                WriteEvent(ResolutionStartEventId, FormatIPAddressNullTerminated(address, stackalloc char[MaxIPFormattedLength]));
+                return ValueStopwatch.StartNew();
+            }
+
+            return default;
+        }
+
+        [NonEvent]
+        public void AfterResolution(string hostNameOrAddress, ValueStopwatch stopwatch, bool successful)
+        {
+            if (stopwatch.IsActive)
+            {
+                double duration = stopwatch.GetElapsedTime().TotalMilliseconds;
+
+                _lookupsDuration!.WriteMetric(duration);
+
+                if (successful)
+                {
+                    ResolutionSuccess(hostNameOrAddress, duration);
+                }
+                else
+                {
+                    ResolutionFailure(hostNameOrAddress, duration);
+                }
+            }
+        }
+
+        [NonEvent]
+        public void AfterResolution(IPAddress address, ValueStopwatch stopwatch, bool successful)
+        {
+            if (stopwatch.IsActive)
+            {
+                double duration = stopwatch.GetElapsedTime().TotalMilliseconds;
+
+                _lookupsDuration!.WriteMetric(duration);
+
+                WriteEvent(
+                    successful ? ResolutionSuccessEventId : ResolutionFailureEventId,
+                    FormatIPAddressNullTerminated(address, stackalloc char[MaxIPFormattedLength]),
+                    duration);
+            }
+        }
+
+
+        [Event(ResolutionSuccessEventId, Level = EventLevel.Informational)]
+        private void ResolutionSuccess(string hostNameOrAddress, double duration) => WriteEvent(ResolutionSuccessEventId, hostNameOrAddress, duration);
+
+        [Event(ResolutionFailureEventId, Level = EventLevel.Informational)]
+        private void ResolutionFailure(string hostNameOrAddress, double duration) => WriteEvent(ResolutionFailureEventId, hostNameOrAddress, duration);
+
+
+        [NonEvent]
+        private static Span<char> FormatIPAddressNullTerminated(IPAddress address, Span<char> destination)
```

Comment:
Do static methods have to be marked with `NonEvent`?

Reply:
You should mark any methods that are not events as `[NonEvent]`, but no, I don't believe that this is strictly required for static methods.